### ISSUE TITLE
CFFI backend bug?

### ISSUE THREAD

**Issue by mingwandroid** [zeromq/pyzmq#1174]
I tried to enable support for testing the CFFI backend over at `Anaconda Distribution`. Testing with:

`PYZMQ_BACKEND=cffi py.test --pyargs zmq.tests.test_socket`

.. fails with:

```
=================================================================================================================================== FAILURES ===================================================================================================================================
___________________________________________________________________________________________________________________________ TestSocket.test_tracker ____________________________________________________________________________________________________________________________

self = <zmq.tests.test_socket.TestSocket testMethod=test_tracker>

    @skip_pypy
    def test_tracker(self):
        "test the MessageTracker object for tracking when zmq is done with a buffer"
        addr = 'tcp://127.0.0.1'
        # get a port:
        sock = socket.socket()
        sock.bind(('127.0.0.1', 0))
        port = sock.getsockname()[1]
        iface = "%s:%i" % (addr, port)
        sock.close()
        time.sleep(0.1)

        a = self.context.socket(zmq.PUSH)
        b = self.context.socket(zmq.PULL)
        self.sockets.extend([a,b])
        a.connect(iface)
        time.sleep(0.1)
        p1 = a.send(b'something', copy=False, track=True)
        assert isinstance(p1, zmq.MessageTracker)
>       assert p1 is zmq._FINISHED_TRACKER
E       AssertionError: assert <zmq.sugar.tracker.MessageTracker object at 0x110e80630> is <zmq.sugar.tracker.MessageTracker object at 0x110def048>
E        +  where <zmq.sugar.tracker.MessageTracker object at 0x110def048> = zmq._FINISHED_TRACKER
```

I tested this on macOS.

**Comment by minrk** [comment]
The `@skip_pypy` decorator on that test should really be skipping if the backend is CFFI, since it's testing functionality only in the Cython backend. While it's *possible* to use the CFFI backend with CPython, there isn't much reason to do so as it's lacking a few features, in particular any form of message tracking.

**Comment by mingwandroid** [comment]
> While it's possible to use the CFFI backend with CPython, there isn't much reason to do so as it's lacking a few features, in particular any form of message tracking.

I figured some projects may want it as a backup in case things go wrong with the `cython` backend and having two backends easily switchable is very useful as a debugging tool in my experience.

Still, I looked deeper into it and enabling the `cffi` backend requires not only `cffi` at *runtime*, but worse, compilers at runtime and fixing that looks like a lot of effort to me.

**Comment by minrk** [comment]
Right, pyzmq doesn't build the cffi backend when installing the Cython backend. It would have to be built at runtime. People are free to use it, but this is not a case that I think is worth supporting at the level of installing the CFFI runtime as part of pyzmq installation, since it is a very rare use case indeed, and not one I would encourage.

If you are building a conda package, though, it might be enough to run:

```
PYZMQ_BACKEND=cffi python -c import zmq
```

once as part of build.sh. I haven't tested this, though.


I'm curious. What prompted investigating this?

**Comment by mingwandroid** [comment]
Oh not much just when investigating bugs having options to enable and disable certain code paths is often useful.

From my entry to the code I didn't get the feeling that it shouldn't be possible to have both backends enabled but that entry involves neither reading docs nor using the code from any code I will ever write.

I did try enabling it that way too and while it does build me the extension module, that module doesn't get used at runtime, instead it attempts to compile it again.

Cheers.
